pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,260 @@
+package nstemplateset
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/codeready-toolchain/toolchain-common/pkg/condition"
+	"github.com/go-logr/logr"
+	"github.com/operator-framework/operator-sdk/pkg/predicate"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	errs "github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("controller_nstemplateset")
+
+const (
+	// Status condition reasons
+	unableToProvisionReason          = "UnableToProvision"
+	unableToProvisionNamespaceReason = "UnableToProvisionNamespace"
+	provisioningNamespaceReason      = "ProvisioningNamespace"
+	provisioningReason               = "Provisioning"
+	provisionedReason                = "Provisioned"
+)
+
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNSTemplateSet{
+		client:        mgr.GetClient(),
+		scheme:        mgr.GetScheme(),
+		applyTemplate: applyTemplate,
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("nstemplateset-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.NSTemplateSet{}}, &handler.EnqueueRequestForObject{}, predicate.GenerationChangedPredicate{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resource
+	enqueueRequestForOwner := &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &toolchainv1alpha1.NSTemplateSet{},
+	}
+	if err := c.Watch(&source.Kind{Type: &corev1.Namespace{}}, enqueueRequestForOwner); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileNSTemplateSet{}
+
+type ReconcileNSTemplateSet struct {
+	client        client.Client
+	scheme        *runtime.Scheme
+	applyTemplate func(client.Client, toolchainv1alpha1.Namespace, map[string]string) error
+}
+
+// Reconcile reads that state of the cluster for a NSTemplateSet object and makes changes based on the state read
+// and what is in the NSTemplateSet.Spec
```

Comment:
Reconcile reads ~~that~~ state of --> Reconcile reads the state of

Reply:
Auto gen comment for all ctrl.  We can keep as it is.